storage/zynq7000-flash/flashdrv.c

```diff
@@ -0,0 +1,498 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Zynq-7000 nor flash driver
+ *
+ * Copyright 2021, 2022 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "qspi.h"
+#include "flashcfg.h"
+
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/time.h>
+
+#define TIMEOUT_CMD_MS 0x05
+#define MAX_SIZE_CMD   0x20
+
+
+struct {
+	/* Buffers for command transactions */
+	uint8_t cmdRx[MAX_SIZE_CMD];
+	uint8_t cmdTx[MAX_SIZE_CMD];
+
+	/* Data caching in sector's buffer */
+	uint8_t *buff;
+	uint32_t buffPos;
+	uint32_t regID;
+	uint32_t sectID;
+
+	flash_info_t info;
+} fdrv_common;
+
+
+/* Auxiliary functions */
+
+static inline time_t flashdrv_timeMsGet(void)
+{
+	time_t now;
+	gettime(&now, NULL);
+
+	return now / 1000;
+}
+
+
+static int flashdrv_regionFind(addr_t offs)
+{
+	int i;
+	size_t sz;
+	addr_t start = 0;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+
+	for (i = 0; i < cfi->regsCount; ++i) {
+		sz = CFI_SIZE_REGION(cfi->regs[i].size, cfi->regs[i].count);
+		if (offs >= start && offs < (start + sz))
+			return i;
+
+		start = sz;
+	}
+
+	return -EINVAL;
+}
+
+
+static size_t flashdrv_regStart(int id)
+{
+	int i;
+	size_t regStart = 0;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+
+	for (i = 0; i < id; ++i)
+		regStart += CFI_SIZE_REGION(cfi->regs[i].size, cfi->regs[i].count);
+
+	return regStart;
+}
+
+
+static inline void flashdrv_initCmdBuff(const flash_cmd_t *cmd, addr_t offs)
+{
+	memset(fdrv_common.cmdTx, 0, cmd->size);
+	fdrv_common.cmdTx[0] = cmd->opCode;
+	fdrv_common.cmdTx[1] = (offs >> 16) & 0xff;
+	fdrv_common.cmdTx[2] = (offs >> 8) & 0xff;
+	fdrv_common.cmdTx[3] = offs & 0xff;
+}
+
+
+/* Auxiliary flash commands */
+
+static int flashdrv_cfiRead(flash_cfi_t *cfi)
+{
+	ssize_t res;
+	size_t cmdSz, dataSz;
+	/* RDID instruction is a generic for each flash memory */
+	const flash_cmd_t cmds = { .opCode = FLASH_CMD_RDID, .size = 1, .dummyCyc = 24 };
+
+	/* Cmd size is rounded to 4 bytes, it includes dummy cycles [b]. */
+	cmdSz = (cmds.size + cmds.dummyCyc / 8 + 0x3) & ~0x3;
+	/* Size of the data which is received during cmd transfer */
+	dataSz = cmds.dummyCyc / 8;
+
+	memset(fdrv_common.cmdTx, 0, cmdSz);
+	fdrv_common.cmdTx[0] = cmds.opCode;
+
+	qspi_start();
+	if ((res = qspi_transfer(fdrv_common.cmdTx, fdrv_common.cmdRx, cmdSz, TIMEOUT_CMD_MS)) < 0) {
```

Comment:
@agkaminski , I believe that we agreed to use the bellow (as example) construct to be MISRA compliant something have changed ?, I ask because I like current approach `if ((res = func()) < 0)` so please clarify.
```C
res = func();
if (res < 0)
```

Reply:
You're right

Reply:
Changed mentioned construction to be compliant with MISRA.